packages/fxa-auth-server/scripts/remove-accounts.ts

```diff
@@ -0,0 +1,175 @@
+import { StripeHelper } from '../lib/payments/stripe';
+import { reportSentryError } from '../lib/sentry';
+const config = require('../config').getProperties();
+
+export async function retreiveUnverifiedAccounts(
+  database: any
+): Promise<Account[]> {
+  const unverifiedAccounts = await database.listAllUnverifiedAccounts();
+  const accountsToDelete: Account[] = [];
+
+  for (const account of unverifiedAccounts) {
+    const accountCreationDate = new Date(account.createdAt);
+    const cutOffDate = new Date();
+    cutOffDate.setDate(cutOffDate.getDate() - 16);
+
+    if (account.verifierSetAt === 0 && accountCreationDate <= cutOffDate) {
+      accountsToDelete.push(account);
+    }
+  }
+
+  return accountsToDelete;
+}
+
+export async function cancelSubscriptionsAndDeleteCustomer(
+  stripeHelper: StripeHelper,
+  paypalHelper: PayPalHelper,
+  account: Account
+): Promise<void> {
+  const stripeCustomer = await stripeHelper.fetchCustomer(account.uid, [
+    'subscriptions',
+  ]);
+
+  if (stripeCustomer) {
+    const paymentProvider = (
+      await stripeHelper.getPaymentProvider(stripeCustomer)
+    ).toLowerCase();
+
+    // Cancel any subscriptions and issue refunds
+    if (stripeCustomer.subscriptions) {
+      for (const subscription of stripeCustomer.subscriptions.data) {
+        await issueRefund(
+          stripeHelper,
+          paypalHelper,
+          subscription,
+          paymentProvider
+        );
+        await stripeHelper.cancelSubscription(subscription.id);
+      }
+    }
+
+    // Remove the customer from stripe
+    await stripeHelper.removeCustomer(account.uid, stripeCustomer.email!);
+
+    // Delete any PayPal billing agreements
+    if (paymentProvider === 'paypal') {
+      await deletePayPalBillingAgreements(paypalHelper, account);
+    }
+  }
+}
+
+export async function issueRefund(
+  stripeHelper: StripeHelper,
+  paypalHelper: PayPalHelper,
+  subscription: Stripe.Subscription,
+  paymentProvider: string
+) {
+  const latestInvoice = await stripeHelper.getInvoice(
+    subscription.latest_invoice as string
+  );
+
+  switch (paymentProvider) {
+    case 'stripe':
+      await stripeHelper.refundPayment(
+        latestInvoice.payment_intent as string,
+        'fraudulent'
+      );
+      break;
+    case 'paypal':
+      const paypalTransactionId =
+        await stripeHelper.getInvoicePaypalTransactionId(latestInvoice);
+      if (paypalTransactionId) {
+        await paypalHelper.issueRefund(latestInvoice, paypalTransactionId);
+      }
+      break;
+    default:
+      break;
+  }
+}
+
+export async function deletePayPalBillingAgreements(
+  paypalHelper: PayPalHelper,
+  account: Account
+) {
+  const paypalAgreements = await getAllPayPalBAByUid(account.uid);
+  const activeIds = paypalAgreements.filter((ba: any) => {
+    ['active', 'pending', 'suspended'].includes(ba.status.toLowerCase());
+  });
+
+  await Promise.allSettled(
```

Comment:
What's the expected behaviour when one of the promises rejects?

Reply:
The difference between `allSettled` and `all` is that the former won't reject if one of the promises rejects or errors. I think this is why we are trying to move away from `all` which does reject if any of the constituent promises fails or errors. For this use case though, I do not think that there will ever be multiple paypal billing agreements because this is for unverified accounts, i mostly lifted this from another place in the app but the methods return arrays of paypal billing agreements. 

My thought on this is that we dont want to block on this because removing the paypall billing agreement is essentially a database clean up step for us. At this point, we will have already issued refunds, canceled subscriptions in stripe, and removed the stripe customer. The last 2 steps after this are similarly important: removing their account from our database and sending out the email. I would be hesitant to say that we should block those last 2 steps on failure here.